pkg/controller/delay.go

```diff
@@ -0,0 +1,252 @@
+	// https://kubernetes.io/docs/concepts/workloads/controllers/statefulset/#stable-network-id
+
+	for index := from; index < to; index++ {
+		ep := endpoint{
+			namespace: namespace,
+			podName:   fmt.Sprintf("%v-%v", serviceName, index),
+			index:     index,
+		}
+
+		ep.url = *url
+		ep.url.Host = fmt.Sprintf("%s.%v.%v.svc.cluster.local", ep.podName, serviceName, ep.namespace)
+
+		eps = append(eps, ep)
+	}
+
+	return eps
+}
+
+func callPrepareDownscaleAndReturnMaxPrepareTimestamp(ctx context.Context, logger log.Logger, client httpClient, endpoints []endpoint) (time.Time, error) {
+	if len(endpoints) == 0 {
+		return time.Now(), nil
+	}
+
+	var (
+		maxTimeMu sync.Mutex
+		maxTime   time.Time
+	)
+
+	type expectedResponse struct {
+		Timestamp int64 `json:"timestamp"`
+	}
+
+	g, ctx := errgroup.WithContext(ctx)
+	for ix := range endpoints {
+		ep := endpoints[ix]
+		g.Go(func() error {
+			target := ep.url.String()
+
+			epLogger := log.With(logger, "pod", ep.podName, "url", target)
+
+			// POST -- prepare for delayed downscale, if not yet prepared, and return timestamp when prepare was called.
+			req, err := http.NewRequestWithContext(ctx, http.MethodPost, target, nil)
+			if err != nil {
+				level.Error(epLogger).Log("msg", "error creating HTTP POST request to endpoint", "err", err)
+				return err
+			}
+
+			resp, err := client.Do(req)
+			if err != nil {
+				level.Error(epLogger).Log("error sending HTTP POST request to endpoint", "err", err)
+				return err
+			}
+
+			defer resp.Body.Close()
+
+			body, readError := io.ReadAll(resp.Body)
+			if readError != nil {
+				level.Error(epLogger).Log("msg", "error reading response from HTTP POST request to endpoint", "err", err)
+				return err
+			}
+
+			if resp.StatusCode/100 != 2 {
+				err := errors.New("HTTP DELETE request returned non-2xx status code")
+				level.Error(epLogger).Log("msg", "unexpected status code returned when calling DELETE on endpoint", "status", resp.StatusCode, "response_body", string(body))
+				return errors.Join(err, readError)
+			}
+
+			r := expectedResponse{}
+			if err := json.Unmarshal(body, &r); err != nil {
+				level.Error(epLogger).Log("msg", "error decoding response from HTTP POST request to endpoint", "err", err)
+				return err
+			}
+
+			if r.Timestamp == 0 {
+				level.Error(epLogger).Log("msg", "invalid response from HTTP POST request to endpoint: no timestamp")
+				return fmt.Errorf("no timestamp in response")
+			}
+
+			t := time.Unix(r.Timestamp, 0)
+
+			maxTimeMu.Lock()
+			if t.After(maxTime) {
+				maxTime = t
+			}
+			maxTimeMu.Unlock()
+
+			level.Debug(epLogger).Log("msg", "HTTP POST request to endpoint succeded", "timestamp", t.UTC().Format(time.RFC3339))
+			return nil
+		})
+	}
+	err := g.Wait()
+	return maxTime, err
+}
+
+func callCancelDelayedDownscale(ctx context.Context, logger log.Logger, client httpClient, endpoints []endpoint) {
+	if len(endpoints) == 0 {
+		return
+	}
+
+	g, _ := errgroup.WithContext(ctx)
```

Comment:
[question, non blocking] Do we want to `SetLimit()` (even an high value) so that we don't spawn a very high number of concurrent calls in a large Mimir cluster?

Reply:
I didn't know about `SetLimit`, nice tip! I've set it to 32.